db/flush.go

```diff
@@ -25,36 +24,10 @@ func (l *Loader) Flush(ctx context.Context, outputModuleHash string, cursor *sin
 			}
 		}
 	}()
-
-	var entryCount int
-	for entriesPair := l.entries.Oldest(); entriesPair != nil; entriesPair = entriesPair.Next() {
-		tableName := entriesPair.Key
-		entries := entriesPair.Value
-
-		if l.tracer.Enabled() {
-			l.logger.Debug("flushing table entries", zap.String("table_name", tableName), zap.Int("entry_count", entries.Len()))
-		}
-
-		for entryPair := entries.Oldest(); entryPair != nil; entryPair = entryPair.Next() {
-			entry := entryPair.Value
-
-			query, err := entry.query(l.getDialect())
-			if err != nil {
-				return fmt.Errorf("failed to get query: %w", err)
-			}
-
-			if l.tracer.Enabled() {
-				l.logger.Debug("adding query from operation to transaction", zap.Stringer("op", entry), zap.String("query", query))
-			}
-
-			if _, err := tx.ExecContext(ctx, query); err != nil {
-				return fmt.Errorf("executing query %q: %w", query, err)
-			}
-		}
-
-		entryCount += entries.Len()
+	entryCount, err := l.getDialect().Flush(tx, ctx, l, outputModuleHash, cursor)
+	if err != nil {
+		return fmt.Errorf("dialect flush: %w", err)
 	}
-
 	entryCount += 1
 	if err := l.UpdateCursor(ctx, tx, outputModuleHash, cursor); err != nil {
 		return fmt.Errorf("update cursor: %w", err)
```

Comment:
This means that `Clickhouse` cursor update is not atomic anymore, 2 first table could be committed and pushed, third one fails.

Net effect, on restart we are going to try to write record that already exists since we will redo 2 first table. 

In `Clickhouse` case, is it a problem?

Reply:
Clickhouse doesn't have transactions so there's no guarantee it is atomic (even if we used the older architecture).
The driver implements transactions as batch inserts.

There are no guarantees of uniqueness so it doesn't throw errors in case of duplicates.

It's better to save twice because you can use: `OPTIMIZE TABLE <table> FINAL DEDUPLICATE` to remove duplicates.